lib/pbench/server/api/__init__.py

```diff
@@ -295,6 +295,16 @@ def post(self):
             )
         md5sum = request.headers.get("md5sum")
 
+        if not request.headers.get("hostname"):
```

Comment:
I'd suggest a more specific extension header for this, like maybe `X-Pbench-Controller-Host` so there's no confusion.

For that matter, we should probably be using the standard `Content-MD5` instead of our own `md5sum`.

Reply:
hmm, I didnt know there is standard header for md5, thats good

Reply:
Also, Flask should allow us to declare endpoints which have variable names in the URL.  For example, today at [line 42](https://github.com/distributed-system-analysis/pbench/blob/73c20444d14ccd03e163dad4fcee25b0edd12081/lib/pbench/server/api/__init__.py#L42) we just declare the `Upload` end-point URL to be `/upload`.  But as we [read in the docs](https://flask-restful.readthedocs.io/en/latest/api.html#flask_restful.Api.add_resource), we can instead declare the resource as:

```
api.add_resource(Upload, f"{app.config['REST_URI']}/upload/ctrl/<controller>")
api.add_resource(Upload, f"{app.config['REST_URI']}/upload/user/<user>")
```

And then the `controller` and `user` variables will be passed when the `post()` method is invoked.  That avoids using another header argument, allowing us to easily track behaviors in the logs.  Further, today we need to use `controller`, but tomorrow we'll be doing this by `user` (the controller information is in the tar ball and will be derived from it instead of required for placement).

Also, we should consider using `PUT` instead of `POST`, as I think we want the idempotency of the operation (see [REST - PUT vs POST](https://restfulapi.net/rest-put-vs-post/)).

Reply:
Should I address the `md5sum` header change comment in this PR or a separate one (I can do it here its a small change but another thought says its different change than this issue)

Reply:
@portante I am bit confused about PUT vs POST, 
the doc says this for PUT

> Use PUT when you want to modify a singular resource which is already a part of resources collection. PUT replaces the resource in its entirety

and for POST

>  Use POST when you want to add a child resource under resources collection.

What we are doing is actually adding a new tarfile under the hostname directory not replacing/ modifying the existing resource/ tarfile. Isnt what we are doing is closer to the POST operation?

Reply:
However, if our resource is controller directory then yes we are modifying the resources then we are closer to the PUT request. I am just confused whether our resource is controller or a tarfile

Reply:
But I think what you said also make sense, we actually PUT a tarfile inside a controller directory

Reply:
@portante interesting read https://stackoverflow.com/questions/56240547/should-http-put-create-a-resource-if-it-does-not-exist. Seems like RFC 2616 is obsolete now

Reply:
okay that make sense

Reply:
I really like the idea of putting the controller (and maybe host) path in the URL.

There's a lot of ambiguity in `PUT` vs `POST` for object creation. Idempotency is definitely an issue, because `PUT` usually specifies the specific resource (which means you already need to know the URI) whereas POST generally creates a resource and returns the URI (e.g., via the `Location` response header... which we may want to consider anyway).

I'm not sure there's really much ground here for "purity" arguments anyway. And there's no reason we can't allow retrying a POST as long as there's a discernible URI from the arguments and payload of the request ... which in this case, I believe we have.

Reply:
Regarding creating a resource, the pbench server semantics today are that the client determines the URI of the resource (albeit, indirectly).  The `controller`/`<tar ball>` name forms the URI.  If two different clients send results today to the pbench server where it happens that they both have controllers with the same name and both provide the same tar ball name, then one will be accepted and one will be quarantined.

I think we should keep the same semantic we have today given we are not changing that functionality with this commit.  To that end, the updated RFCs recommends `PUT` for situations where the resource URI is known ahead of time.

Further, both [OpenStack Swift] (https://docs.openstack.org/api-ref/object-store/) and [Amazon S3](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html) object stores allow for PUT operations.  They also both support `POST` operations, OpenStack Swift with slight semantic differences, and S3 considering it just another way to send the same data.

We could decide to enhance this PR to support both mechanisms to do the same thing.

However, I'd rather only support `PUT` today, so that we can support for `POST` later if we decide we want to take away control of the URI construction from the client when the proper back-end infrastructure plumbed out.